unstructured/ingest/v2/processes/connectors/salesforce.py

```diff
@@ -0,0 +1,293 @@
+from textwrap import dedent
+from typing import TYPE_CHECKING, Any, Generator
+
+from dateutil import parser
+
+from unstructured.documents.elements import DataSourceMetadata
+from unstructured.ingest.enhanced_dataclass import enhanced_field
+from unstructured.ingest.error import SourceConnectionNetworkError
+from unstructured.ingest.v2.interfaces import (
+    AccessConfig,
+    ConnectionConfig,
+    Downloader,
+    DownloaderConfig,
+    DownloadResponse,
+    FileData,
+    Indexer,
+    IndexerConfig,
+    SourceIdentifiers,
+)
+from unstructured.ingest.v2.logger import logger
+from unstructured.ingest.v2.processes.connector_registry import (
+    SourceRegistryEntry,
+    add_source_entry,
+)
+from unstructured.utils import requires_dependencies
+
+
+class MissingCategoryError(Exception):
+    """There are no categories with that name."""
+
+
+CONNECTOR_TYPE = "salesforce"
+
+if TYPE_CHECKING:
+    pass
+
+SALESFORCE_API_VERSION = "57.0"
+
+ACCEPTED_CATEGORIES = ["Account", "Case", "Campaign", "EmailMessage", "Lead"]
+
+EMAIL_TEMPLATE = Template(
+    """MIME-Version: 1.0
+Date: $date
+Message-ID: $message_identifier
+Subject: $subject
+From: $from_email
+To: $to_email
+Content-Type: multipart/alternative; boundary="00000000000095c9b205eff92630"
+--00000000000095c9b205eff92630
+Content-Type: text/plain; charset="UTF-8"
+$textbody
+--00000000000095c9b205eff92630
+Content-Type: text/html; charset="UTF-8"
+$htmlbody
+--00000000000095c9b205eff92630--
+""",
+)
+
+
+@dataclass
+class SalesforceAccessConfig(AccessConfig):
+    consumer_key: str
+    private_key: str
+
+    @requires_dependencies(["cryptography"])
+    def get_private_key_value_and_type(self) -> t.Tuple[str, t.Type]:
+        from cryptography.hazmat.primitives import serialization
+
+        try:
+            serialization.load_pem_private_key(data=self.private_key.encode("utf-8"), password=None)
+        except ValueError:
+            pass
+        else:
+            return self.private_key, str
+
+        if Path(self.private_key).is_file():
+            return self.private_key, Path
+
+        raise ValueError("private_key does not contain PEM private key or path")
+
+
+@dataclass
+class SalesforceConnectionConfig(ConnectionConfig):
+    username: str
+    access_config: SalesforceAccessConfig = enhanced_field(sensitive=True)
+
+    @requires_dependencies(["simple_salesforce"], extras="salesforce")
+    def get_client(self):
+        from simple_salesforce import Salesforce
+
+        pkey_value, pkey_type = self.access_config.get_private_key_value_and_type()
+
+        return Salesforce(
+            username=self.username,
+            consumer_key=self.access_config.consumer_key,
+            privatekey_file=pkey_value if pkey_type is Path else None,
+            privatekey=pkey_value if pkey_type is str else None,
+            version=SALESFORCE_API_VERSION,
+        )
+
+
+@dataclass
+class SalesforceIndexerConfig(IndexerConfig):
+    categories: t.List[str]
```

Comment:
This might be better as a set of Literals ([[Dataclass argument choices with a default option](https://stackoverflow.com/questions/61756716/dataclass-argument-choices-with-a-default-option)](https://stackoverflow.com/questions/61756716/dataclass-argument-choices-with-a-default-option))
```suggestion
    categories: set[Literal[ACCEPTED_CATEGORIES]]
```

Not sure if this works but could be a nice check on the dataclass itself.

Reply:
@rbiseck3 This does seem to work as `Literal[ACCEPTED_CATEGORIES]` it didn't like `set[Literal[ACCEPTED_CATEGORIES]]`

Reply:
hmmm. Now it doesn't work anymore.